src/Illuminate/Foundation/Testing/Concerns/InteractsWithDatabase.php

```diff
@@ -26,6 +26,14 @@ trait InteractsWithDatabase
      */
     protected function assertDatabaseHas($table, array $data, $connection = null)
     {
+        if ($table instanceof Model) {
+            foreach ($data as $key => $value) {
+                $table->setAttribute($key, $value);
```

Comment:
This will have unintended side effects of setting the given model's attributes to the provided `$data` parameter. 

To accomplish this without making breaking changes to the model's underlying methods to access casting would probably be to create a new instance of the model and force fill it:

```php
if ($table instanceof Model) {
    $data = (new $table)->forceFill($data)->getAttributes();
}
```